File: frontend/src/api/tracking.ts
import client from "./client";

/* ── Response types matching backend schemas ── */

export interface DailySkinScoreResponse {
  score_id: string;
  user_id: string;
  measurement_id: string | null;
  score_date: string; // YYYY-MM-DD
  overall_score: number;
  hydration_norm: number | null;
  elasticity_norm: number | null;
  pore_norm: number | null;
  wrinkle_norm: number | null;
  pigmentation_norm: number | null;
  trend_direction: string | null;
  trend_velocity: number | null;
  ma_7_score: number | null;
  is_anomaly: boolean;
  created_at: string;
}

export interface TrendResponse {
  trend_direction: "improving" | "stable" | "declining";
  trend_velocity: number;
  recent_avg: number | null;
  data_points: number;
}

export interface SummaryResponse {
  period: string;
  from_date: string;
  to_date: string;
  avg_score: number | null;
  max_score: number | null;
  min_score: number | null;
  measurement_count: number;
  metric_averages: {
    hydration: number | null;
    elasticity: number | null;
    pore: number | null;
    wrinkle: number | null;
    pigmentation: number | null;
  };
}

export interface DetectionSignalsResponse {
  date: string;
  improvement: {
    detected: boolean;
    improved_metrics?: Array<{
      metric: string;
      before_mean: number;
      after_mean: number;
      change_pct: number;
    }>;
  };
  deterioration: {
    level: number;
    watch?: { deteriorated_metrics: Array<{ metric: string }> };
    warning?: { deteriorated_metrics: Array<{ metric: string }> };
    alert?: { deteriorated_metrics: Array<{ metric: string }> };
  };
  direction: "improved" | "stable" | "deteriorated";
  product_impacts: Array<{
    product_id: string;
    product_name: string;
    impact_probability: number;
    direction: string;
    affected_metrics: string[];
  }>;
}

/* ── API functions ── */

export async function getScores(params: {
  period?: "daily" | "weekly" | "monthly";
  from?: string;
  to?: string;
}): Promise<DailySkinScoreResponse[]> {
  const { data } = await client.get<DailySkinScoreResponse[]>(
    "/v1/tracking/scores",
    { params },
  );
  return data;
}

export async function getTrend(): Promise<TrendResponse> {
  const { data } = await client.get<TrendResponse>("/v1/tracking/trend");
  return data;
}

export async function getSummary(params?: {
  period?: string;
  from?: string;
  to?: string;
}): Promise<SummaryResponse> {
  const { data } = await client.get<SummaryResponse>("/v1/tracking/summary", {
    params,
  });
  return data;
}

export async function getDetectionSignals(
  targetDate?: string,
): Promise<DetectionSignalsResponse> {
  const { data } = await client.get<DetectionSignalsResponse>(
    "/v1/detection/signals",
    { params: targetDate ? { target_date: targetDate } : undefined },
  );
  return data;
}

File: backend/app/schemas/tracking.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class DailySkinScoreCreate(BaseModel):
    user_id: str
    measurement_id: str | None = None
    score_date: date
    overall_score: int = Field(..., ge=0, le=100)
    hydration_norm: float | None = Field(None, ge=0, le=1)
    elasticity_norm: float | None = Field(None, ge=0, le=1)
    pore_norm: float | None = Field(None, ge=0, le=1)
    wrinkle_norm: float | None = Field(None, ge=0, le=1)
    pigmentation_norm: float | None = Field(None, ge=0, le=1)
    trend_direction: Literal["improving", "stable", "declining"] | None = None
    trend_velocity: float | None = None
    ma_7_score: float | None = None
    is_anomaly: bool = False


class DailySkinScoreRead(BaseModel):
    score_id: str
    user_id: str
    measurement_id: str | None = None
    score_date: date
    overall_score: int
    hydration_norm: float | None = None
    elasticity_norm: float | None = None
    pore_norm: float | None = None
    wrinkle_norm: float | None = None
    pigmentation_norm: float | None = None
    trend_direction: str | None = None
    trend_velocity: float | None = None
    ma_7_score: float | None = None
    is_anomaly: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class ProductEffectAnalysisCreate(BaseModel):
    user_id: str
    product_id: str
    analysis_period_start: date
    analysis_period_end: date
    effect_score: float = Field(..., ge=-100, le=100)
    metric_deltas: dict
    confidence_level: float = Field(..., ge=0, le=1)
    usage_duration_days: int
    before_avg_score: float | None = None
    after_avg_score: float | None = None
    sample_count: int
    analysis_version: str = "1.0"


class ProductEffectAnalysisRead(BaseModel):
    analysis_id: str
    user_id: str
    product_id: str
    analysis_period_start: date
    analysis_period_end: date
    effect_score: float
    metric_deltas: dict
    confidence_level: float
    usage_duration_days: int
    before_avg_score: float | None = None
    after_avg_score: float | None = None
    sample_count: int
    analysis_version: str
    created_at: datetime

    model_config = {"from_attributes": True}
